kliff/trainer/base_trainer.py

```diff
@@ -0,0 +1,636 @@
+        implement. Although to ensure that `initialize` function remains consistent
+        this method will not raise NotImplemented error, rather it will quietly pass.
+        So be aware.
+        """
+        pass
+
+    def setup_optimizer(self):
+        """
+        Set up the optimizer based on the provided information. If the optimizer is not
+        provided, it will be loaded from the optimizer_name. If the optimizer_name is not
+        provided, it will raise an error. If the optimizer_provider is scipy, it will be
+        loaded from the scipy.optimize. If the optimizer_provider is torch, it will be
+        loaded from the torch.optim. Left for the derived classes to implement.
+        """
+        raise TrainerError("setup_optimizer not implemented.")
+
+    def setup_test_train_datasets(self):
+        """
+        Simple test train split for now, will have more options like stratification
+         in the future.
+
+        """
+        # test train splits
+        train_size = self.training_manifest.get("training_dataset", {}).get(
+            "train_size", len(self.dataset)
+        )
+        val_size = self.training_manifest.get("validation_dataset", {}).get(
+            "val_size", 0
+        )
+
+        # sanity checks
+        if not isinstance(train_size, int) or train_size < 1:
+            logger.warning(
+                "Train size is not provided or is less than 1. Using full dataset for training."
+            )
+            train_size = len(self.dataset)
+        else:
+            logger.info(f"Training dataset size: {train_size}")
+
+        if not isinstance(val_size, int) or val_size < 0:
+            logger.warning(
+                "Val size is not provided or is less than 0. Using 0 for validation."
+            )
+            val_size = 0
+        else:
+            logger.info(f"Validation dataset size: {val_size}")
+
+        if train_size + val_size > len(self.dataset):
+            raise TrainerError(
+                "Sum of train, val, and test sizes is greater than the dataset size."
+            )
+
+        # check if indices are provided
+        train_indices = self.dataset_sample_manifest.get("train_indices")
+        if train_indices is None:
+            train_indices = np.arange(train_size, dtype=int)
+        elif isinstance(train_indices, str):
+            train_indices = np.genfromtxt(train_indices, dtype=int)
+        else:
+            TrainerError("train_indices should be a numpy array or a path to a file.")
+
+        val_indices = self.dataset_sample_manifest.get("val_indices")
+        if val_indices is None:
+            val_indices = np.arange(train_size, train_size + val_size, dtype=int)
+        elif isinstance(val_indices, str):
+            val_indices = np.genfromtxt(val_indices, dtype=int)
+        else:
+            TrainerError("val_indices should be a numpy array or a path to a file.")
+
+        if self.dataset_manifest.get("shuffle", False):
```

Comment:
I don't think `shuffle` here has any major effect. We will need to shuffle at each training epoch if using mini-batch. Instead, if the entire training set is used to compute the gradients, no shuffle is needed.

Reply:
Agree. Removed shuffling.